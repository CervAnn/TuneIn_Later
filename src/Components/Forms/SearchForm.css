* {
    margin: 0;
    box-sizing: border-box;
}

button {
    font-family: 'Permanent Marker', cursive;
    background-color: transparent;
    border: none;
    background-position: center;
}
