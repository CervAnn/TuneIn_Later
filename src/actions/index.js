export const addAlbums = (albums) => ({
	type: 'ADD_ALBUMS',
	albums
});

export const validUser = (user) => ({
	type: 'VALID_USER',
	user
});

export const addFavorite = (id) => ({
	type: 'ADD_FAVORITE',
	id
});

export const invalidUser = () => ({
	type: 'INVALID_USER'
});

export const getUserFavorites = (userFavorites) => ({
	type: 'GET_USER_FAVORITES',
	userFavorites
});