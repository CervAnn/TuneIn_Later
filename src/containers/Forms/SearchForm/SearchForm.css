* {
    box-sizing: border-box;
    font-family: 'Permanent Marker', cursive;
    margin: 0;
}

#search-button {
    background-color: transparent;
    background-position: center;
    border: none;
    color: #F4E697;
    font-size: 30px;
    width: auto;
}

#search-button:hover {
    border-radius: 15px;
    border: 2px solid #cc7b3a;
    width: 100%;
  }

