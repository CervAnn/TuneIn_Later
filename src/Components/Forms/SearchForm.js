import React, { Component } from 'react';
import { fetchAlbums } from '../../util/apiCalls';
import { addAlbums } from '../../actions/index';
import { connect } from 'react-redux';

class SearchForm extends Component {
	constructor() {
		super();
		this.state = {
				artist: ""
		}
	}

	handleChange = (e) => {
		this.setState({[e.target.name]: e.target.value})
	}

	handleSubmit = (e) => {
		e.preventDefault()
		const queryArtist = {
				name: this.state.artist
		}
		this.searchArtist(queryArtist)
		this.clearInputs()
	}

	clearInputs = () => {
		this.setState({artist: ""})
	}

	searchArtist = (queryArtist) => {
		fetchAlbums(queryArtist)
		.then(albums => this.props.addAlbums(albums.results))
		.catch(error => console.log(error))
	}

	render() {
	console.log(this.props.addAlbums)
		return (
			<form className="SearchForm">
					<input
					type = "text"
					placeholder = "Artist:"
					name = "artist"
					value = {this.state.artist}
					onChange = {this.handleChange}
					/>
					<button onClick={event => this.handleSubmit(event)}>Search</button>
				<button>Logout</button>
			</form>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
  addAlbums: (albums) => dispatch(addAlbums(albums))
});

export default connect(null, mapDispatchToProps)(SearchForm);
