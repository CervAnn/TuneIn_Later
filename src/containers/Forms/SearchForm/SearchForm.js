import React, { Component } from "react";
import { fetchAlbums } from "../../../util/apiCalls";
import { addAlbums } from "../../../actions/index";
import { connect } from "react-redux";
import "./SearchForm.css";
import PropTypes from 'prop-types';

export class SearchForm extends Component {
  constructor() {
    super();
    this.state = {
      artist: "",
      error: ""
    };
  }

  handleChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  handleSubmit = e => {
    e.preventDefault();
    const queryArtist = {
      name: this.state.artist
    };
    this.searchArtist(queryArtist);
    this.clearInputs();
  };

  clearInputs = () => {
    this.setState({ artist: "", error: "" });
  };

  searchArtist = queryArtist => {
    fetchAlbums(queryArtist)
      .then(albums => this.props.addAlbums(albums.results))
      .catch(error => this.setState({ error: error }));
  };

  render() {
    return (
      <section className="SearchFormContainer">
        {this.state.error && (
          <p id="error">
            There was an issue retrieving your artist's albums. Please try
            again.
          </p>
        )}
        <form id="SearchForm">
          <input
            type="text"
            placeholder="Artist:"
            name="artist"
            value={this.state.artist}
            onChange={this.handleChange}
          />
          <button
            id="search-button"
            onClick={event => this.handleSubmit(event)}
          >
            Search
          </button>
        </form>
      </section>
    );
  }
}

export const mapStateToProps = state => ({
  user: state.user
});

export const mapDispatchToProps = dispatch => ({
  addAlbums: albums => dispatch(addAlbums(albums))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchForm);

SearchForm.propTypes = {
  user: PropTypes.object,
  mapDispatchToProps: PropTypes.func,
  mapStateToProps: PropTypes.func,
}