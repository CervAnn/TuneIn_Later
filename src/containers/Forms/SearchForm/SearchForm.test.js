import React from 'react';
import { shallow } from 'enzyme';
import { SearchForm, mapStateToProps, mapDispatchToProps } from './SearchForm';
import { addAlbums } from '../../../actions/index'

describe('SearchForm', () => {
  let wrapper, mockUser, mockAddAlbums;

  beforeEach(() => {
    mockUser = null;
    mockAddAlbums = jest.fn();
    wrapper = shallow(<SearchForm 
    user={mockUser}
    addAlbums={mockAddAlbums}
    />);
  });

  it('should match the snapshot with all data passed in correctly', () => {

    expect(wrapper).toMatchSnapshot();
  });



  it('should update state on a change in an input', () => {

    const mockEvent = { target: { name: 'artist', value: 'Beyonce' } }

    wrapper.find('input').at(0).simulate('change', mockEvent);

    expect(wrapper.state('artist')).toEqual('Beyonce')
  });

  it('should call handleChange on a keydown event', () => {

    const mockEvent = () => {}

    wrapper.instance().handleChange = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('input').simulate('change', mockEvent);

  expect(wrapper.instance().handleChange).toHaveBeenCalled();
  });

  it('should call handleSubmit on a button click', () => {

    const mockEvent = {target: {name: 'artist', value: 'Beyonce'}}

    wrapper.instance().handleSubmit = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('button').simulate('click', mockEvent);

    expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
  });


  it('should update the state when clearInputs is called', () => {

    const expected = '';

    wrapper.setState({
      artist: 'Beyonce',
      error: 'There was an error.'
    });

    wrapper.instance().clearInputs();

    expect(wrapper.state('artist')).toEqual(expected);
    expect(wrapper.state('error')).toEqual(expected);
  });

  it('should run searchArtist on a button click', () => {
    const mockEvent = { preventDefault: jest.fn() }

    wrapper.instance().searchArtist = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('button').simulate('click', mockEvent);

    expect(wrapper.instance().searchArtist).toHaveBeenCalled();
  });

  describe('mapStateToProps', () => {

    it('should return an object with the current user', () => {
      const mockState = {
        albums:[],
        user: { id: 1, name: 'Alan', email: 'alan@turing.io' },
        error: '',
        favorites: [
          {
            album_id: 626204707,
            artist_name: 'BeyoncÃ©',
            album_name: '4 (Expanded Edition)',
            price: 11.99,
            artwork_url: 'https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg',
            release_date: '2013-12-13T08:00:00Z',
            content_advisory_rating: 'Explicit',
            primary_genre_name: 'Pop',
          }
        ]
      }
  
      const expected = {
        user: { id: 1, name: 'Alan', email: 'alan@turing.io' }
      }
  
      const mappedProps = mapStateToProps(mockState);
  
      expect(mappedProps).toEqual(expected)
    });

  });

  describe('mapDispatchToProps', () => {

    it('should dispatch addAlbums when called', () => {
      const mockAlbums = { album_name: "Homecoming", artist_name: 'Beyonce', album_id: 12345 }
      const mockDispatch = jest.fn();
      const actionToDispatch = addAlbums(mockAlbums);
    
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addAlbums({ album_name: "Homecoming", artist_name: 'Beyonce', album_id: 12345 });
    
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

  });

});