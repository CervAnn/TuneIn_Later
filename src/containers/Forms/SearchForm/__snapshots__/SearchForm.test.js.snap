// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot with all data passed in correctly 1`] = `
<section
  className="SearchFormContainer"
>
  <form
    id="SearchForm"
  >
    <input
      name="artist"
      onChange={[Function]}
      placeholder="Artist:"
      type="text"
      value=""
    />
    <button
      id="search-button"
      onClick={[Function]}
    >
      Search
    </button>
  </form>
</section>
`;
